shapely/geometry/point.py

```diff
@@ -95,9 +95,10 @@ def y(self):
     @property
     def z(self):
         """Return z coordinate."""
-        if not shapely.has_z(self):
+        z = shapely.get_z(self)
+        if np.isnan(z):
```

Comment:
```suggestion
        if np.isnan(z) and not shapely.has_z(self):
```

This would ensure we keep returning NaN if it actually has a Z coord that is NaN, while for the common case avoiding the `has_z` check

Reply:
That makes sense to me. Will be faster for all use cases except for those where nan was passed in as a variable to the Point. I think that is a pretty good trade off.